// @feezal/feezal-element-basic-template/feezal-element-basic-template.js
import {FeezalElement, html} from "@feezal/feezal-element";

class FeezalElementBasicTemplate extends FeezalElement {
    static get template() {
        return html`
            <style include="feezal-style-element"></style>
            <div id="content"></div>
        `;
    }
    static get properties() {
        return {
            topic: {
                type: String,
                value: '',
                reflectToAttribute: true,
                observer: '_topicChanged'
            },
            content: {
                type: String,
                value: '',
                reflectToAttribute: true,
                observer: '_contentChanged',
                feezal: {
                    textarea: true
                }
            },
            hideUndefined: {
                type: Boolean,
                value: true,
                reflectToAttribute: true,
                observer: '_contentChanged',
            },
            msg: {
                type: Object,
                value: {},
                observer: '_contentChanged'
            }
        }
    }
    _processTemplate(template, vars = {}) {
        const match = template.match(/\${[^}]+}/g);
        if (match) {
            match.forEach(v => {
                const key = v.substr(2, v.length - 3);
                const rx = new RegExp('\\${' + key + '}', 'g');
                let replacement;
                if (this.hideUndefined && typeof vars[key] === 'undefined') {
                    replacement = '';
                } else {
                    replacement = vars[key];
                }
                template = template.replace(rx, replacement);
            });
        }
        return template;
    }
    _contentChanged() {
        this.$.content.innerHTML = this._processTemplate(this.content, this.msg);
    }

    static get feezal() {
        return {
            palette: {
                category: 'Basic',
                name: 'Template',
                color: '#eee'
            },
            attributes: [
                'content'
            ],
            styles: [
                'top',
                'left',
                'width',
                'height',
                'font',
                'color',
                'text-align',
                'background',
                'border',
                'overflow',
            ],
            restrict: {
                minWidth: 12,
                minHeight: 12
            },
            defaultStyle: {
                width: '60px',
                height: '20px'
            }
        };
    }

    connectedCallback() {
        super.connectedCallback();
        feezal.connection.subscribe(this.topic, msg => {
            this.msg = msg;
        });
    }
    _topicChanged() {

    }
}

window.customElements.define('feezal-element-basic-template', FeezalElementBasicTemplate);

export {FeezalElementBasicTemplate};
