// @feezal/feezal-element-basic-template/feezal-element-basic-template.js
import {FeezalElement, html} from "@feezal/feezal-element";

class FeezalElementBasicTemplate extends FeezalElement {
    static get template() {
        return html`
            <style include="feezal-style-element"></style>
            <div id="content"></div>
        `;
    }
    static get properties() {
        return {
            topic: {
                type: String,
                value: '',
                reflectToAttribute: true,
                observer: '_topicChanged'
            },
            msg: {
                type: Object,
                value: {},
                observer: '_msgChanged'
            }
        }
    }

    static get feezal() {
        return {
            palette: {
                category: 'Basic',
                name: 'Template',
                color: '#eee'
            },
            attributes: [
                'topic',
                {name: 'template', textarea: true, template: true}
            ],
            styles: [
                'top',
                'left',
                'width',
                'height',
                'font',
                'color',
                'text-align',
                'background',
                'border',
                'overflow',
            ],
            restrict: {
                minWidth: 12,
                minHeight: 12
            },
            defaultStyle: {
                width: '60px',
                height: '20px',
                color: 'var(--primary-text-color)'
            }
        };
    }

    _msgChanged() {
        if (!this._processTemplate && this.querySelector('template')) {
            // evil eval...
            this._processTemplate = new Function('msg', "return `" + this.querySelector('template').innerHTML + "`;");
        }

        if (!this.msg || Object.keys(this.msg).length === 0) {
            return;
        }
        try {
            this.$.content.innerHTML = this._processTemplate(this.msg);
        } catch (error) {
            console.error(error.message);
        }
    }

    _topicChanged() {
        if (this.subscription) {
            feezal.connection.unsubscribe(this.subscription);
        }
        if (this.topic) {
            this.subscription = feezal.connection.subscribe(this.topic, msg => this.msg = msg);
        }
    }
}

window.customElements.define('feezal-element-basic-template', FeezalElementBasicTemplate);

export {FeezalElementBasicTemplate};
