// @feezal/feezal-element-basic-view/feezal-element-basic-view.js
import {FeezalElement, html} from "@feezal/feezal-element";

class FeezalElementBasicView extends FeezalElement {
    static get template() {
        return html`
            <style include="feezal-style-element"></style>
            <style>
                :host {
                    overflow: hidden;
                }
                #content {
                    width: 100%;
                    height: 100%;
                }
                #content.editor {
                    background-image:
                        -moz-linear-gradient(45deg, #eee 25%, transparent 25%,transparent 75%, #eee 75%, #eee 100%),
                        -moz-linear-gradient(45deg, #eee 25%, transparent 25%,transparent 75%, #eee 75%, #eee 100%);
                    
                    background-image:
                        -webkit-linear-gradient(45deg, #eee 25%, transparent 25%,transparent 75%, #eee 75%, #eee 100%),
                        -webkit-linear-gradient(45deg, #eee 25%, transparent 25%,transparent 75%, #eee 75%, #eee 100%);
                    
                    -moz-background-size:20px 20px;
                    background-size:20px 20px;
                    -webkit-background-size:21px 21px;
                    background-position:0 0, 10px 10px;
                    display: flex;
                }
            </style>
            <div id="content"></div>
        `;
    }
    static get properties() {
        return {
            subscribeTopic: {
                type: String,
                value: '',
                reflectToAttribute: true
            },
            view: {
                type: String,
                value: '',
                reflectToAttribute: true,
                observer: '_viewChanged'
            }
        }
    }

    static get feezal() {
        return {
            palette: {
                category: 'Basic',
                name: 'View',
                color: '#eee'
            },
            attributes: [
                {
                    name: 'view',
                    dropdown: 'views'
                }
            ],
            baseAttribute: 'view',
            styles: [
                'top',
                'left',
                'width',
                'height',
                'font',
                'color',
                'text-align',
                'background',
                'border',
                'overflow',
            ],
            restrict: {
                minWidth: 12,
                minHeight: 12
            },
            defaultStyle: {
                width: '60px',
                height: '20px'
            }
        };
    }
    connectedCallback() {
        super.connectedCallback();
        if (feezal.isEditor) {
            this.$.content.classList.add('editor');
        }
    }
    _viewChanged() {
        if (feezal.isEditor) {
            this.$.content.innerHTML = '<span style="margin: auto; font-weight: bold;">View: ' + this.view + '</span>';
        } else {
            this.$.content.innerHTML = feezal.site.querySelector('feezal-view[name="' + this.view + '"]').outerHTML;
        }
    }
}

window.customElements.define('feezal-element-basic-view', FeezalElementBasicView);

export {FeezalElementBasicView};
