// @feezal/feezal-element-paper-slider/feezal-element-paper-slider.js
import {html} from '@polymer/polymer/polymer-element.js';
import {FeezalElement} from "@feezal/feezal-element";

import '@polymer/paper-slider';

class FeezalElementPaperSlider extends FeezalElement {
    static get template() {
        return html`
            <style include="feezal-style-element"></style>
            <style>
                :host {
                    overflow: visible;
                }
                paper-slider {
                    width: 100%;
                    height: 100%;
                }
            </style>
            <paper-slider id="slider"
                noink="[[noink]]" 
                pin="[[pin]]" 
                disabled="[[disabled]]" 
                min="[[min]]"
                max="[[max]]"
            >[[label]]</paper-slider>
        `;
    }
    static get properties() {
        return {
            subscribeTopic: {
                type: String,
                value: '',
                reflectToAttribute: true
            },
            publishTopic: {
                type: String,
                value: '',
                reflectToAttribute: true
            },
            min: {
                type: Number,
                value: 0,
                reflectToAttribute: true
            },
            max: {
                type: Number,
                value: 100,
                reflectToAttribute: true
            },
            disabled: {
                type: Boolean,
                value: false,
                reflectToAttribute: true
            },
            immediate: {
                type: Boolean,
                value: false,
                reflectToAttribute: true
            },
            immediateThrottle: {
                type: Number,
                value: 100,
                reflectToAttribute: true
            },
            noink: {
                type: Boolean,
                value: false,
                reflectToAttribute: true
            },
            pin: {
                type: Boolean,
                value: true,
                reflectToAttribute: true
            }
        }
    }
    static get feezal() {
        return {
            palette: {
                category: 'Paper',
                name: 'Slider',
                color: '#eee'
            },
            attributes: [
                {name: 'min', size: 'half'},
                {name: 'max', size: 'half'},
                {name: 'immediate', size: 'half'},
                {name: 'immediateThrottle', size: 'half'},
                {name: 'pin'},
                {name: 'noink'},
                {name: 'disabled'},
            ],
            baseAttribute: 'value',
            styles: [
                'top',
                'left',
                'width',
                'height',
                {'property': '--paper-slider-container-color', type: 'color'},
                {'property': '--paper-slider-bar-color', type: 'color'},
                {'property': '--paper-slider-active-color', type: 'color'},
                {'property': '--paper-slider-secondary-color', type: 'color'},
                {'property': '--paper-slider-knob-color', type: 'color'},
                {'property': '--paper-slider-disabled-knob-color', type: 'color'},
                {'property': '--paper-slider-pin-color', type: 'color'},
                {'property': '--paper-slider-font-color', type: 'color'},
                {'property': '--paper-slider-markers-color', type: 'color'},
                {'property': '--paper-slider-disabled-active-color', type: 'color'},
                {'property': '--paper-slider-disabled-secondary-color', type: 'color'},
                {'property': '--paper-slider-knob-start-color', type: 'color'},
                {'property': '--paper-slider-knob-start-border-color', type: 'color'},
                {'property': '--paper-slider-pin-start-color', type: 'color'},
                {'property': '--paper-slider-height'}
            ],
            restrict: {
                minWidth: 100,
                minHeight: 22
            },
            defaultStyle: {
                width: '100px',
                height: '22px'
            }
        }
    }

    connectedCallback() {
        super.connectedCallback();
        if (this.immediate) {
            this.$.slider.addEventListener('immediate-value-change', this.throttle(() => {
                feezal.connection.publish(this.publishTopic, this.$.slider.immediateValue);
            }, this.immediateThrottle));
        }
        this.$.slider.addEventListener('change', () => {
            feezal.connection.publish(this.publishTopic, this.$.slider.value);
        });
        this.subscribe(this.subscribeTopic, msg => {
            if (!this.$.slider.dragging && !this._inThrottle) {
                this.$.slider.setAttribute('value', Number(msg.payload));
            }
        });
    }
}

window.customElements.define('feezal-element-paper-slider', FeezalElementPaperSlider);

export {FeezalElementPaperSlider};
