// @feezal/feezal-element-material-light/feezal-element-material-light.js
/* global feezal */
import {FeezalElement, feezalBaseStyles, html, css} from '@feezal/feezal-element';
import {svg, LitElement} from 'lit';
import '@material/web/slider/slider.js';
import '@material/web/select/outlined-select.js';
import '@material/web/select/select-option.js';

// ─── Arc geometry (0° = top, clockwise, matches feezal-element-material-gauge convention) ───
const CX = 50;
const CY = 50;
const TRACK_R = 40;   // brightness ring centre-line radius
const RING_W  = 7;    // ring stroke width
const CENTER_R = 26;  // centre-circle radius (toggle / colour zone)
const POWER_R  = 10;  // inner power-toggle zone radius (within centre)
const ARC_START = 225;
const ARC_SWEEP = 270;

function polarXY(deg, r) {
    const rad = (deg - 90) * Math.PI / 180;
    return [+(CX + r * Math.cos(rad)).toFixed(3), +(CY + r * Math.sin(rad)).toFixed(3)];
}

function arcPath(fromDeg, toDeg, r) {
    const [ax, ay] = polarXY(fromDeg, r);
    const [bx, by] = polarXY(toDeg, r);
    const sweep = ((toDeg - fromDeg) + 360) % 360;
    return `M${ax},${ay} A${r},${r} 0 ${sweep > 180 ? 1 : 0},1 ${bx},${by}`;
}

function pctToAngle(pct) {
    return (ARC_START + (Math.max(0, Math.min(100, pct)) / 100) * ARC_SWEEP) % 360;
}

function angleToPct(deg) {
    const arcEnd = (ARC_START + ARC_SWEEP) % 360; // 135
    let n;
    if (deg >= ARC_START) n = deg - ARC_START;
    else if (deg <= arcEnd) n = deg + (360 - ARC_START);
    else n = deg < 180 ? ARC_SWEEP : 0; // gap zone — clamp to nearest end
    return Math.max(0, Math.min(100, Math.round(n / ARC_SWEEP * 100)));
}

// ─── Colour helpers ───────────────────────────────────────────────────────────
function kelvinToRgb(k) {
    // Simple warm-to-cool interpolation: 2700 K (warm) → 6500 K (cool)
    const t = Math.max(0, Math.min(1, (k - 2700) / 3800));
    return [Math.round(255 - t * 60), Math.round(210 + t * 45), Math.round(90 + t * 165)];
}

function rgbToHex(r, g, b) {
    return '#' + [r, g, b].map(c => Math.round(Math.max(0, Math.min(255, c))).toString(16).padStart(2, '0')).join('');
}

function rgbToHsv(r, g, b) {
    r /= 255; g /= 255; b /= 255;
    const max = Math.max(r, g, b), d = max - Math.min(r, g, b);
    let h = 0;
    if (d) {
        if (max === r) h = ((g - b) / d + (g < b ? 6 : 0)) * 60;
        else if (max === g) h = ((b - r) / d + 2) * 60;
        else h = ((r - g) / d + 4) * 60;
    }
    return {h, s: max ? d / max : 0, v: max};
}

function hsvToRgb(h, s, v) {
    const i = Math.floor(h / 60) % 6;
    const f = h / 60 - Math.floor(h / 60);
    const p = v * (1 - s), q = v * (1 - f * s), t = v * (1 - (1 - f) * s);
    const combos = [[v, t, p], [q, v, p], [p, v, t], [p, q, v], [t, p, v], [v, p, q]];
    const [r, g, b] = combos[i];
    return [Math.round(r * 255), Math.round(g * 255), Math.round(b * 255)];
}

function parseRgb(raw) {
    try {
        if (typeof raw === 'string' && !raw.startsWith('[')) {
            const parts = raw.split(',').map(Number);
            if (parts.length >= 3 && parts.every(n => !isNaN(n))) return parts.slice(0, 3);
        }
        const arr = typeof raw === 'string' ? JSON.parse(raw) : raw;
        if (Array.isArray(arr)) return arr.slice(0, 3).map(Number);
    } catch {}
    return null;
}

// CIE 1931 xy (+ optional brightness) → sRGB. Used to display the colour of
// zigbee2mqtt / HA lights that report colour as {x, y} (e.g. Philips Hue).
function xyToRgb(x, y, bri = 1) {
    if (!y) return [255, 255, 255];
    const Y = bri;
    const X = (Y / y) * x;
    const Z = (Y / y) * (1 - x - y);
    let r =  X * 1.656492 - Y * 0.354851 - Z * 0.255038;
    let g = -X * 0.707196 + Y * 1.655397 + Z * 0.036152;
    let b =  X * 0.051713 - Y * 0.121364 + Z * 1.011530;
    const gamma = c => (c <= 0.0031308 ? 12.92 * c : 1.055 * Math.pow(c, 1 / 2.4) - 0.055);
    [r, g, b] = [r, g, b].map(c => gamma(Math.max(0, c)));
    const max = Math.max(r, g, b, 1);
    return [r, g, b].map(c => Math.round(Math.max(0, Math.min(1, c / max)) * 255));
}

// ─── Element ──────────────────────────────────────────────────────────────────
class FeezalElementMaterialLight extends FeezalElement {
    static get feezal() {
        return {
            palette: {name: 'Light', category: 'Device', color: '#1565c0', icon: 'lightbulb'},
            description: 'Smart light card — brightness ring, colour temperature, RGB/HS colour wheel, white channel, and effect selector.',
            // ── N6 custom inspector (E35) ──────────────────────────────────
            inspector: 'feezal-element-material-light-inspector',
            // ── N12 Auto-Discovery descriptor (E35 extended) ───────────────
            // zigbee2mqtt / HA discovery always emits the JSON schema, so an
            // auto-configured light defaults to payload-mode: json with a single
            // subscribe (base topic) / publish (…/set) pair. Capability ranges
            // (brightness scale, mired→kelvin temp range, effect list) and the
            // active colour control are mapped from the config too.
            discovery: {
                component: 'light',
                map: {
                    // wiring — schema decides separate vs json
                    schema:        {attr: 'payload-mode', valueMap: {json: 'json', _default: 'separate'}},
                    state_topic:   {attr: 'subscribe', onlyWhen: {schema: 'json'}}, // base topic
                    command_topic: {attr: 'publish',   onlyWhen: {schema: 'json'}}, // …/set
                    // capability ranges
                    brightness_scale:      {attr: 'brightness-max'},          // e.g. 254 → 0–100 %
                    supported_color_modes: {attr: 'mode', transform: 'colorMode'},
                    min_mireds: {attr: 'color-temp-max', unit: 'mired\u2192kelvin'}, // 153 → 6536 K
                    max_mireds: {attr: 'color-temp-min', unit: 'mired\u2192kelvin'}, // 500 → 2000 K
                    // effects
                    effect_list: {attr: 'effects', transform: 'join'},
                    // availability (a single small "unavailable" badge on the card)
                    availability_topic: {attr: 'subscribe-availability'},
                    // device name → label
                    name: 'label',
                },
            },
            attributes: [
                // Wiring mode
                {name: 'payload-mode', type: 'select', options: ['separate', 'json'], default: 'separate', help: 'separate = one topic per property; json = single topic carrying a JSON object.'},
                {name: 'subscribe', type: 'mqttTopic', help: 'json mode: base topic carrying the whole JSON state object.'},
                {name: 'publish',   type: 'mqttTopic', help: 'json mode: command topic (usually …/set) that accepts a partial JSON object.'},
                {name: 'json-map',  type: 'string', default: '', help: 'json mode: optional JSON string overriding the default property→key map.'},
                // On/off
                {name: 'subscribe-state',   type: 'mqttTopic', help: 'Topic receiving on/off state.'},
                {name: 'publish-state',     type: 'mqttTopic', help: 'Topic to publish on/off.'},
                {name: 'payload-on',        type: 'string',    default: 'on',  help: 'Payload representing "on".'},
                {name: 'payload-off',       type: 'string',    default: 'off', help: 'Payload representing "off".'},
                // Brightness
                {name: 'subscribe-brightness', type: 'mqttTopic', help: 'Current brightness (0–100 %).'},
                {name: 'publish-brightness',   type: 'mqttTopic', help: 'Publish brightness on ring release.'},
                {name: 'brightness-min', type: 'number', default: 0,   size: 'half', help: 'Minimum brightness value on the MQTT topic.'},
                {name: 'brightness-max', type: 'number', default: 100, size: 'half', help: 'Maximum brightness value on the MQTT topic.'},
                // Colour temperature
                {name: 'subscribe-color-temp', type: 'mqttTopic', help: 'Current colour temperature.'},
                {name: 'publish-color-temp',   type: 'mqttTopic', help: 'Publish colour temperature.'},
                {name: 'color-temp-unit', type: 'select', options: ['kelvin', 'mired'], default: 'kelvin', help: 'Unit used on colour-temp topics.'},
                {name: 'color-temp-min',  type: 'number', default: 2700, help: 'Minimum colour temperature (K).'},
                {name: 'color-temp-max',  type: 'number', default: 6500, help: 'Maximum colour temperature (K).'},
                // RGB / HS
                {name: 'subscribe-rgb', type: 'mqttTopic', help: 'Current RGB value (JSON [r,g,b] or "r,g,b").'},
                {name: 'publish-rgb',   type: 'mqttTopic', help: 'Publish RGB value as JSON [r,g,b].'},
                {name: 'subscribe-hs',  type: 'mqttTopic', help: 'Current hue/saturation (JSON [h,s]).'},
                {name: 'publish-hs',    type: 'mqttTopic', help: 'Publish hue/saturation as JSON [h,s].'},
                // Mode
                {name: 'mode', type: 'select', options: ['brightness', 'color_temp', 'rgb', 'hs'], default: 'brightness', help: 'Control mode shown in the card centre.'},
                // Effects
                {name: 'subscribe-effect', type: 'mqttTopic', help: 'Current effect name.'},
                {name: 'publish-effect',   type: 'mqttTopic', help: 'Publish selected effect name.'},
                {name: 'effects',          type: 'string',    default: '', help: 'Comma-separated list of available effect names.'},
                // White / RGBW / RGBWW
                {name: 'subscribe-white',      type: 'mqttTopic', help: 'White channel (0–100 %) for RGBW lamps.'},
                {name: 'publish-white',        type: 'mqttTopic', help: 'Publish white channel.'},
                {name: 'subscribe-warm-white', type: 'mqttTopic', help: 'Warm-white channel (0–100 %) for RGBWW lamps.'},
                {name: 'publish-warm-white',   type: 'mqttTopic', help: 'Publish warm-white channel.'},
                {name: 'subscribe-cold-white', type: 'mqttTopic', help: 'Cold-white channel (0–100 %) for RGBWW lamps.'},
                {name: 'publish-cold-white',   type: 'mqttTopic', help: 'Publish cold-white channel.'},
                // Availability
                {name: 'subscribe-availability', type: 'mqttTopic', help: 'Topic reporting device availability. When unavailable a small badge is shown; the control stays usable.'},
                {name: 'payload-available',      type: 'string', default: 'online',  help: 'Payload meaning the device is available.'},
                {name: 'payload-unavailable',    type: 'string', default: 'offline', help: 'Payload meaning the device is unavailable.'},
                // Label
                {name: 'label', type: 'string', default: '', help: 'Optional label shown below the circle.'}
            ],
            styles: [
                'top', 'left', 'width', 'height', 'background', 'border-radius',
                // Theme-aware colour tokens — editable in the Style inspector.
                // The `default` is shown as the field placeholder so the active
                // mapping is visible; leave a field blank to inherit it.
                // on-color  → filled brightness arc, drag knob, centre tint/outline while ON
                // off-color → unfilled ring groove + centre outline while OFF
                {property: '--feezal-light-on-color',      type: 'color', default: 'var(--primary-text-color)'},
                {property: '--feezal-light-off-color',     type: 'color', default: 'var(--secondary-text-color)'},
                {property: '--feezal-light-surface-color', type: 'color', default: 'var(--primary-background-color)'},
                {property: '--feezal-light-text-color',    type: 'color', default: 'var(--primary-text-color)'},
                {property: '--feezal-light-error-color',   type: 'color', default: 'var(--error-color)'}
            ],
            restrict: {minWidth: 120, minHeight: 140},
            defaultStyle: {width: '180px', height: '220px'}
        };
    }

    static properties = {
        payloadMode:        {type: String, reflect: true, attribute: 'payload-mode'},
        publish:            {type: String, reflect: true, attribute: 'publish'},
        jsonMap:            {type: String, reflect: true, attribute: 'json-map'},
        subscribeState:     {type: String, reflect: true, attribute: 'subscribe-state'},
        publishState:       {type: String, reflect: true, attribute: 'publish-state'},
        payloadOn:          {type: String, reflect: true, attribute: 'payload-on'},
        payloadOff:         {type: String, reflect: true, attribute: 'payload-off'},
        subscribeBrightness:{type: String, reflect: true, attribute: 'subscribe-brightness'},
        publishBrightness:  {type: String, reflect: true, attribute: 'publish-brightness'},
        brightnessMin:      {type: Number, reflect: true, attribute: 'brightness-min'},
        brightnessMax:      {type: Number, reflect: true, attribute: 'brightness-max'},
        subscribeColorTemp: {type: String, reflect: true, attribute: 'subscribe-color-temp'},
        publishColorTemp:   {type: String, reflect: true, attribute: 'publish-color-temp'},
        colorTempUnit:      {type: String, reflect: true, attribute: 'color-temp-unit'},
        colorTempMin:       {type: Number, reflect: true, attribute: 'color-temp-min'},
        colorTempMax:       {type: Number, reflect: true, attribute: 'color-temp-max'},
        subscribeRgb:       {type: String, reflect: true, attribute: 'subscribe-rgb'},
        publishRgb:         {type: String, reflect: true, attribute: 'publish-rgb'},
        subscribeHs:        {type: String, reflect: true, attribute: 'subscribe-hs'},
        publishHs:          {type: String, reflect: true, attribute: 'publish-hs'},
        mode:               {type: String, reflect: true},
        subscribeEffect:    {type: String, reflect: true, attribute: 'subscribe-effect'},
        publishEffect:      {type: String, reflect: true, attribute: 'publish-effect'},
        effects:            {type: String, reflect: true},
        subscribeWhite:     {type: String, reflect: true, attribute: 'subscribe-white'},
        publishWhite:       {type: String, reflect: true, attribute: 'publish-white'},
        subscribeWarmWhite: {type: String, reflect: true, attribute: 'subscribe-warm-white'},
        publishWarmWhite:   {type: String, reflect: true, attribute: 'publish-warm-white'},
        subscribeColdWhite: {type: String, reflect: true, attribute: 'subscribe-cold-white'},
        publishColdWhite:   {type: String, reflect: true, attribute: 'publish-cold-white'},
        subscribeAvailability: {type: String, reflect: true, attribute: 'subscribe-availability'},
        payloadAvailable:   {type: String, reflect: true, attribute: 'payload-available'},
        payloadUnavailable: {type: String, reflect: true, attribute: 'payload-unavailable'},
        discoveryId:        {type: String, reflect: true, attribute: 'discovery-id'},
        label:              {type: String, reflect: true},
        // Internal state
        _on:        {state: true},
        _brt:       {state: true},   // brightness 0–100
        _colorTemp: {state: true},   // colour temperature in Kelvin
        _rgb:       {state: true},   // [r, g, b]
        _hs:        {state: true},   // [h, s]  h: 0–360, s: 0–100
        _effect:    {state: true},
        _white:     {state: true},
        _warmWhite: {state: true},
        _coldWhite: {state: true},
        _dragBrt:   {state: true},   // live brightness during ring drag (null = not dragging)
        _available: {state: true},   // device availability (true unless told otherwise)
    };

    static styles = [feezalBaseStyles, css`
        :host {
            display: flex;
            flex-direction: column;
            align-items: center;
            padding: 6px;
            box-sizing: border-box;
            overflow: hidden;
            gap: 4px;
            position: relative;

            /* ── Theme-aware colour tokens (state-aware) ────────────────────
               Five overridable colours, each defaulting to a feezal/HA theme
               variable (with a literal fallback). Override per-element via the
               Style inspector or a theme rule.
                 on-color  → filled brightness arc + drag knob + centre
                             tint/outline while the lamp is ON
                 off-color → unfilled ring groove + centre outline while OFF
               In colour-temperature / RGB / HS modes the ring and centre show
               the live light colour, which overrides on-color. The CT gradient
               and hue wheel stay fixed — they represent the real light output. */
            --feezal-light-on-color:      var(--primary-text-color, var(--primary-color, var(--sl-color-primary-600, #0284c7)));
            --feezal-light-off-color:     var(--secondary-text-color, var(--divider-color, var(--feezal-border, #e0e0e0)));
            --feezal-light-surface-color: var(--primary-background-color, var(--feezal-bg, #fff));
            --feezal-light-text-color:    var(--primary-text-color, var(--feezal-color, #333));
            --feezal-light-error-color:   var(--error-color, #b00020);

            --md-sys-color-primary:    var(--feezal-light-on-color);
            --md-sys-color-surface:    var(--feezal-light-surface-color);
            --md-sys-color-on-surface: var(--feezal-light-text-color);
            /* Outlined-select field + dropdown menu theming (effect selector).
               The menu surface, outline, label and selected-item colours all
               read from the same theme tokens so the popup respects the theme. */
            --md-sys-color-on-surface-variant:        var(--feezal-light-text-color);
            --md-sys-color-outline:                   var(--feezal-light-off-color);
            --md-sys-color-outline-variant:           var(--feezal-light-off-color);
            --md-sys-color-surface-container:         var(--feezal-light-surface-color);
            --md-sys-color-surface-container-high:    var(--feezal-light-surface-color);
            --md-sys-color-surface-container-highest: var(--feezal-light-surface-color);
            --md-sys-color-secondary-container:       var(--feezal-light-on-color);
            --md-sys-color-on-secondary-container:    var(--feezal-light-surface-color);
        }
        .unavail {
            position: absolute;
            top: 4px;
            right: 4px;
            width: 18px;
            height: 18px;
            color: var(--feezal-light-error-color);
            opacity: 0.8;
            pointer-events: none;
            z-index: 2;
        }
        .unavail svg { width: 100%; height: 100%; display: block; }
        .ring-wrap { width: 100%; flex-shrink: 0; }
        svg {
            width: 100%;
            display: block;
            aspect-ratio: 1;
            overflow: visible;
            touch-action: none;
            user-select: none;
        }
        .controls { width: 100%; display: flex; flex-direction: column; gap: 4px; }
        .ctrl-label {
            font-size: 10px; opacity: 0.55;
            color: var(--feezal-light-text-color);
            margin-bottom: -2px;
        }
        .ct-track {
            position: relative; height: 22px; border-radius: 11px;
            background: linear-gradient(to right, #ff8c00, #fff8ee 50%, #aac4ff);
            cursor: pointer; touch-action: none;
        }
        .ct-thumb {
            position: absolute; top: 50%;
            transform: translate(-50%, -50%);
            width: 18px; height: 18px; border-radius: 50%;
            border: 2px solid rgba(0,0,0,0.2);
            background: var(--feezal-light-surface-color); pointer-events: none;
            box-shadow: 0 1px 4px rgba(0,0,0,0.25);
        }
        md-slider { width: 100%; }
        md-outlined-select { width: 100%; }
        .label {
            font-size: 11px; opacity: 0.65; text-align: center;
            color: var(--feezal-light-text-color);
            white-space: nowrap; overflow: hidden; text-overflow: ellipsis; width: 100%;
        }
    `];

    constructor() {
        super();
        this.payloadMode         = 'separate';
        this.publish             = '';
        this.jsonMap             = '';
        this.subscribeState      = '';
        this.publishState        = '';
        this.payloadOn           = 'on';
        this.payloadOff          = 'off';
        this.subscribeBrightness = '';
        this.publishBrightness   = '';
        this.brightnessMin       = 0;
        this.brightnessMax       = 100;
        this.subscribeColorTemp  = '';
        this.publishColorTemp    = '';
        this.colorTempUnit       = 'kelvin';
        this.colorTempMin        = 2700;
        this.colorTempMax        = 6500;
        this.subscribeRgb        = '';
        this.publishRgb          = '';
        this.subscribeHs         = '';
        this.publishHs           = '';
        this.mode                = 'brightness';
        this.subscribeEffect     = '';
        this.publishEffect       = '';
        this.effects             = '';
        this.subscribeWhite      = '';
        this.publishWhite        = '';
        this.subscribeWarmWhite  = '';
        this.publishWarmWhite    = '';
        this.subscribeColdWhite  = '';
        this.publishColdWhite    = '';
        this.subscribeAvailability = '';
        this.payloadAvailable    = 'online';
        this.payloadUnavailable  = 'offline';
        this.discoveryId         = '';
        this.label               = '';
        this._on        = false;
        this._brt       = null;
        this._colorTemp = null;
        this._rgb       = null;
        this._hs        = null;
        this._effect    = '';
        this._white     = null;
        this._warmWhite = null;
        this._coldWhite = null;
        this._dragBrt   = null;
        this._available = true;
        // Non-reactive drag flags
        this.__ctDragging = false;
    }

    // The light fully manages its own subscriptions below; suppress the generic
    // base subscription path (which would otherwise fire for the json-mode
    // `subscribe` topic and try to set attributes by topic suffix).
    _subscribe() { /* intentionally empty — see connectedCallback */ }

    connectedCallback() {
        super.connectedCallback();
        if (feezal.isEditor) return;

        // Availability — always honoured, in both payload modes.
        if (this.subscribeAvailability) {
            this.addSubscription(this.subscribeAvailability, msg => {
                const v = this.getProperty(msg, this.messageProperty);
                const s = String(v).toLowerCase();
                this._available = v === this.payloadAvailable ||
                    (s !== String(this.payloadUnavailable).toLowerCase() &&
                     s !== 'offline' && s !== 'false' && s !== '0' && s !== 'unavailable');
            });
        }

        if (this.payloadMode === 'json') {
            if (this.subscribe) {
                this.addSubscription(this.subscribe, msg => {
                    let obj = this.getProperty(msg, this.messageProperty);
                    if (typeof obj === 'string') {
                        try { obj = JSON.parse(obj); } catch { return; }
                    }
                    if (obj && typeof obj === 'object') this._applyJsonState(obj);
                });
            }
            return;
        }

        // ── Separate (per-topic) mode ──────────────────────────────────────
        if (this.subscribeState) {
            this.addSubscription(this.subscribeState, msg => {
                const v = this.getProperty(msg, this.messageProperty);
                this._on = v === this.payloadOn || v === true || v === 1 || v === '1' ||
                           (typeof v === 'string' && v.toLowerCase() === 'on');
            });
        }
        if (this.subscribeBrightness) {
            this.addSubscription(this.subscribeBrightness, msg => {
                const v = Number(this.getProperty(msg, this.messageProperty));
                if (!isNaN(v)) {
                    const min = this.brightnessMin ?? 0;
                    const max = this.brightnessMax ?? 100;
                    this._brt = max === min ? 0 : Math.max(0, Math.min(100, (v - min) / (max - min) * 100));
                }
            });
        }
        if (this.subscribeColorTemp) {
            this.addSubscription(this.subscribeColorTemp, msg => {
                let v = Number(this.getProperty(msg, this.messageProperty));
                if (!isNaN(v)) {
                    if (this.colorTempUnit === 'mired') v = Math.round(1_000_000 / v);
                    this._colorTemp = v;
                }
            });
        }
        if (this.subscribeRgb) {
            this.addSubscription(this.subscribeRgb, msg => {
                const rgb = parseRgb(this.getProperty(msg, this.messageProperty));
                if (rgb) this._rgb = rgb;
            });
        }
        if (this.subscribeHs) {
            this.addSubscription(this.subscribeHs, msg => {
                try {
                    const raw = this.getProperty(msg, this.messageProperty);
                    const arr = typeof raw === 'string' ? JSON.parse(raw) : raw;
                    if (Array.isArray(arr) && arr.length >= 2) this._hs = arr.slice(0, 2).map(Number);
                } catch {}
            });
        }
        if (this.subscribeEffect) {
            this.addSubscription(this.subscribeEffect, msg => {
                this._effect = String(this.getProperty(msg, this.messageProperty) ?? '');
            });
        }
        const mkSub = (stateProp, topic) => {
            if (topic) {
                this.addSubscription(topic, msg => {
                    const v = Number(this.getProperty(msg, this.messageProperty));
                    if (!isNaN(v)) this[stateProp] = Math.max(0, Math.min(100, v));
                });
            }
        };
        mkSub('_white',     this.subscribeWhite);
        mkSub('_warmWhite', this.subscribeWarmWhite);
        mkSub('_coldWhite', this.subscribeColdWhite);
    }

    // ─── JSON payload mode helpers ────────────────────────────────────────────
    get _jsonMap() {
        const defaults = {
            state: 'state', brightness: 'brightness',
            color_mode: 'color_mode', color_temp: 'color_temp',
            color: 'color', effect: 'effect',
        };
        if (this.jsonMap) {
            try { return {...defaults, ...JSON.parse(this.jsonMap)}; } catch { /* fall through */ }
        }
        return defaults;
    }

    _applyJsonState(obj) {
        const map = this._jsonMap;
        const get = key => this.getProperty(obj, key);

        const state = get(map.state);
        if (state !== undefined && state !== null) {
            const s = String(state).toLowerCase();
            this._on = state === this.payloadOn || state === true || state === 1 ||
                       s === 'on' || s === 'true' || s === '1';
        }

        const bri = Number(get(map.brightness));
        if (!isNaN(bri)) {
            const max = this.brightnessMax || 100;
            this._brt = Math.max(0, Math.min(100, (bri / max) * 100));
        }

        let ct = Number(get(map.color_temp));
        if (!isNaN(ct)) {
            if (this.colorTempUnit === 'mired') ct = Math.round(1_000_000 / ct);
            this._colorTemp = ct;
        }

        const color = get(map.color);
        if (color && typeof color === 'object') {
            if (color.hue !== undefined || color.h !== undefined) {
                this._hs = [Number(color.hue ?? color.h), Number(color.saturation ?? color.s ?? 100)];
            } else if (color.r !== undefined) {
                this._rgb = [Number(color.r), Number(color.g), Number(color.b)];
            } else if (color.x !== undefined && color.y !== undefined) {
                this._rgb = xyToRgb(Number(color.x), Number(color.y));
            }
        }

        const effect = get(map.effect);
        if (effect !== undefined && effect !== null) this._effect = String(effect);
    }

    // Publish a property: in json mode merges into a single JSON object on the
    // `publish` topic; in separate mode publishes `value` to `topic`.
    _pub(topic, value, jsonObj) {
        if (this.payloadMode === 'json') {
            if (this.publish) feezal.connection.pub(this.publish, JSON.stringify(jsonObj));
        } else if (topic) {
            feezal.connection.pub(topic, value);
        }
    }

    // ─── SVG pointer handling ─────────────────────────────────────────────────
    _onSvgPointerDown(e) {
        if (feezal.isEditor) return;
        const {sx, sy} = this._toSvgCoords(e);
        const dx = sx - CX, dy = sy - CY;
        const dist = Math.sqrt(dx * dx + dy * dy);

        if (dist <= CENTER_R) {
            this._handleCenterTap(dx, dy, dist);
        } else if (this._on && dist <= TRACK_R + RING_W) {
            this._startBrtDrag(e, sx, sy);
        }
    }

    _handleCenterTap(dx, dy, dist) {
        const mode = this.mode || 'brightness';
        const isColorMode = mode === 'rgb' || mode === 'hs';
        if (!this._on || !isColorMode || dist <= POWER_R) {
            // Power zone, simple modes, or light is off → toggle
            this._toggle();
        } else {
            // Outer wheel zone in rgb/hs mode while on → colour pick
            this._pickColor(dx, dy, dist);
        }
    }

    _toggle() {
        this._on = !this._on;
        const payload = this._on ? this.payloadOn : this.payloadOff;
        this._pub(this.publishState, payload, {[this._jsonMap.state]: payload});
    }

    _pickColor(dx, dy, dist) {
        // Angle measured clockwise from top (matches the colour wheel segments drawn with polarXY)
        const hue = ((Math.atan2(dy, dx) * 180 / Math.PI) + 90 + 360) % 360;
        const sat = Math.min(1, dist / CENTER_R);
        if (this.mode === 'rgb') {
            const rgb = hsvToRgb(hue, sat, 1);
            this._rgb = rgb;
            this._pub(this.publishRgb, JSON.stringify(rgb),
                {[this._jsonMap.color]: {r: rgb[0], g: rgb[1], b: rgb[2]}});
        } else if (this.mode === 'hs') {
            const hs = [Math.round(hue), Math.round(sat * 100)];
            this._hs = hs;
            this._pub(this.publishHs, JSON.stringify(hs),
                {[this._jsonMap.color]: {hue: hs[0], saturation: hs[1]}});
        }
    }

    _startBrtDrag(e, sx, sy) {
        this._dragBrt = this._brt ?? 0;

        const onMove = ev => {
            const {sx: x, sy: y} = this._toSvgCoords(ev);
            const deg = ((Math.atan2(y - CY, x - CX) * 180 / Math.PI) + 90 + 360) % 360;
            this._dragBrt = angleToPct(deg);
        };
        const onUp = () => {
            document.removeEventListener('pointermove', onMove);
            document.removeEventListener('pointerup', onUp);
            const final = this._dragBrt;
            this._dragBrt = null;
            this._brt = final;
            const min = this.brightnessMin ?? 0;
            const max = this.brightnessMax ?? 100;
            const raw = Math.round(min + (final / 100) * (max - min));
            const jsonRaw = Math.round((final / 100) * (this.brightnessMax || 100));
            this._pub(this.publishBrightness, String(raw), {[this._jsonMap.brightness]: jsonRaw});
        };
        document.addEventListener('pointermove', onMove);
        document.addEventListener('pointerup', onUp);
        // Immediately update to click position
        const deg = ((Math.atan2(sy - CY, sx - CX) * 180 / Math.PI) + 90 + 360) % 360;
        this._dragBrt = angleToPct(deg);
    }

    _toSvgCoords(e) {
        const el = this.shadowRoot.querySelector('svg');
        if (!el) return {sx: 0, sy: 0};
        const r = el.getBoundingClientRect();
        return {sx: ((e.clientX - r.left) / r.width) * 100, sy: ((e.clientY - r.top) / r.height) * 100};
    }

    // ─── Colour temperature track ─────────────────────────────────────────────
    _onCtDown(e) {
        if (feezal.isEditor) return;
        e.currentTarget.setPointerCapture(e.pointerId);
        this.__ctDragging = true;
        this._applyCt(e, false);
    }

    _onCtMove(e) {
        if (!this.__ctDragging) return;
        this._applyCt(e, false);
    }

    _onCtUp(e) {
        if (!this.__ctDragging) return;
        this.__ctDragging = false;
        this._applyCt(e, true);
    }

    _applyCt(e, publish) {
        const rect = e.currentTarget.getBoundingClientRect();
        const ratio = Math.max(0, Math.min(1, (e.clientX - rect.left) / rect.width));
        const min = this.colorTempMin || 2700;
        const max = this.colorTempMax || 6500;
        const ct = Math.round(min + ratio * (max - min));
        this._colorTemp = ct;
        if (publish) {
            const v = this.colorTempUnit === 'mired' ? Math.round(1_000_000 / ct) : ct;
            this._pub(this.publishColorTemp, String(v), {[this._jsonMap.color_temp]: v});
        }
    }

    // ─── White sliders ────────────────────────────────────────────────────────
    _onWhite(e, pubTopic, stateProp) {
        const v = Number(e.target.value);
        this[stateProp] = v;
        if (pubTopic) feezal.connection.pub(pubTopic, String(v));
    }

    // ─── Effect selector ──────────────────────────────────────────────────────
    _onEffect(e) {
        this._effect = e.target.value;
        this._pub(this.publishEffect, this._effect, {[this._jsonMap.effect]: this._effect});
    }

    // ─── Computed values ──────────────────────────────────────────────────────
    get _dispBrt() {
        return this._dragBrt !== null ? this._dragBrt : (this._brt ?? 0);
    }

    get _accentColor() {
        if (!this._on) return 'var(--feezal-light-off-color)';
        const mode = this.mode || 'brightness';
        if (mode === 'rgb' && this._rgb) return rgbToHex(...this._rgb);
        if (mode === 'hs'  && this._hs)  return rgbToHex(...hsvToRgb(this._hs[0], this._hs[1] / 100, 1));
        if (mode === 'color_temp' && this._colorTemp) return rgbToHex(...kelvinToRgb(this._colorTemp));
        return 'var(--feezal-light-on-color)';
    }

    // ─── SVG content ─────────────────────────────────────────────────────────
    _svgContent() {
        const isOn   = feezal.isEditor ? true : this._on;
        const brt    = feezal.isEditor ? 60   : this._dispBrt;
        const accent = feezal.isEditor ? 'var(--feezal-light-on-color)' : this._accentColor;
        const trackC = 'var(--feezal-light-off-color)';
        const mode   = this.mode || 'brightness';

        const arcEndAngle = (ARC_START + ARC_SWEEP) % 360; // 135°
        const fillAngle   = pctToAngle(brt);
        const trackD      = arcPath(ARC_START, arcEndAngle, TRACK_R);
        const fillD       = brt > 0.5 ? arcPath(ARC_START, fillAngle, TRACK_R) : null;
        const [hx, hy]    = polarXY(brt > 0.5 ? fillAngle : ARC_START, TRACK_R);

        return svg`
            <!-- Background track arc -->
            <path d="${trackD}" fill="none" stroke="${trackC}"
                stroke-width="${RING_W}" stroke-linecap="round" pointer-events="none"/>

            <!-- Brightness fill arc -->
            ${isOn && fillD ? svg`
                <path d="${fillD}" fill="none" stroke="${accent}"
                    stroke-width="${RING_W}" stroke-linecap="round" pointer-events="none"/>
            ` : ''}

            <!-- Centre disc -->
            <circle cx="${CX}" cy="${CY}" r="${CENTER_R}"
                fill="var(--feezal-light-surface-color)"
                stroke="${isOn ? accent : trackC}" stroke-width="1.5"
                pointer-events="none"/>

            <!-- On-state tint (visible in brightness mode; CT/colour modes draw over it) -->
            ${isOn ? svg`
                <circle cx="${CX}" cy="${CY}" r="${CENTER_R - 1.5}"
                    fill="${accent}" opacity="0.14" pointer-events="none"/>
            ` : ''}

            <!-- Centre content -->
            ${isOn
                ? this._svgCenter(mode, brt, accent)
                : svg`<text x="${CX}" y="${CY}" text-anchor="middle"
                        dominant-baseline="middle" font-size="9"
                        opacity="0.55" fill="var(--feezal-light-off-color)"
                        pointer-events="none">off</text>`}

            <!-- Drag handle on ring (shown when on) -->
            ${isOn ? svg`
                <circle cx="${hx}" cy="${hy}" r="5"
                    fill="${accent}" stroke="var(--feezal-light-surface-color)" stroke-width="2"
                    pointer-events="none"/>
            ` : ''}`;
    }

    _svgCenter(mode, brt, accent) {
        switch (mode) {
            case 'brightness':
                return svg`
                    <text x="${CX}" y="${CY}" text-anchor="middle"
                        dominant-baseline="middle" font-size="13" font-weight="500"
                        fill="var(--feezal-light-text-color)" pointer-events="none">
                        ${brt !== null ? `${brt}%` : '—'}
                    </text>`;

            case 'color_temp': {
                const k = feezal.isEditor ? 4000 : (this._colorTemp ?? null);
                return svg`
                    <defs>
                        <linearGradient id="feezal-ct-grad" gradientUnits="userSpaceOnUse"
                            x1="${CX - CENTER_R}" y1="${CY}" x2="${CX + CENTER_R}" y2="${CY}">
                            <stop offset="0%"   stop-color="#ff8c00"/>
                            <stop offset="50%"  stop-color="#fff8ee"/>
                            <stop offset="100%" stop-color="#aac4ff"/>
                        </linearGradient>
                        <clipPath id="feezal-ct-clip">
                            <circle cx="${CX}" cy="${CY}" r="${CENTER_R - 1.5}"/>
                        </clipPath>
                    </defs>
                    <circle cx="${CX}" cy="${CY}" r="${CENTER_R - 1.5}"
                        fill="url(#feezal-ct-grad)" clip-path="url(#feezal-ct-clip)"
                        pointer-events="none"/>
                    ${k ? svg`
                        <text x="${CX}" y="${CY}" text-anchor="middle"
                            dominant-baseline="middle" font-size="8.5"
                            fill="rgba(0,0,0,0.5)" pointer-events="none">${k}K</text>
                    ` : ''}`;
            }

            case 'rgb':
            case 'hs': {
                // Hue colour wheel: 36 pie segments (hue = segment angle in polarXY convention)
                const segs = [];
                const N = 36;
                for (let i = 0; i < N; i++) {
                    const a1 = i * (360 / N), a2 = (i + 1) * (360 / N);
                    const [x1, y1] = polarXY(a1, CENTER_R - 1.5);
                    const [x2, y2] = polarXY(a2, CENTER_R - 1.5);
                    segs.push(svg`<path
                        d="M${CX},${CY} L${x1},${y1} A${CENTER_R - 1.5},${CENTER_R - 1.5} 0 0,1 ${x2},${y2} Z"
                        fill="hsl(${a1},100%,50%)" stroke="none" pointer-events="none"/>`);
                }

                // Selected colour indicator dot (computed from current _rgb or _hs)
                let dotX = null, dotY = null;
                if (!feezal.isEditor) {
                    if (mode === 'rgb' && this._rgb) {
                        const {h, s} = rgbToHsv(...this._rgb);
                        // h matches polarXY angle (both: 0=top=red, clockwise)
                        const [px, py] = polarXY(h, s * (CENTER_R - 5));
                        dotX = px; dotY = py;
                    } else if (mode === 'hs' && this._hs) {
                        const [px, py] = polarXY(this._hs[0], (this._hs[1] / 100) * (CENTER_R - 5));
                        dotX = px; dotY = py;
                    }
                }

                return svg`
                    <defs>
                        <clipPath id="feezal-wheel-clip">
                            <circle cx="${CX}" cy="${CY}" r="${CENTER_R - 1.5}"/>
                        </clipPath>
                        <radialGradient id="feezal-sat-grad" cx="50%" cy="50%" r="50%">
                            <stop offset="0%"  stop-color="white" stop-opacity="0.88"/>
                            <stop offset="70%" stop-color="white" stop-opacity="0"/>
                        </radialGradient>
                    </defs>
                    <!-- Hue segments -->
                    <g clip-path="url(#feezal-wheel-clip)" pointer-events="none">${segs}</g>
                    <!-- Saturation overlay (white-to-transparent radial gradient) -->
                    <circle cx="${CX}" cy="${CY}" r="${CENTER_R - 1.5}"
                        fill="url(#feezal-sat-grad)" clip-path="url(#feezal-wheel-clip)"
                        pointer-events="none"/>
                    <!-- Inner power-toggle zone -->
                    <circle cx="${CX}" cy="${CY}" r="${POWER_R}"
                        fill="rgba(0,0,0,0.3)" pointer-events="none"/>
                    <text x="${CX}" y="${CY + 0.5}" text-anchor="middle"
                        dominant-baseline="middle" font-size="11"
                        fill="white" pointer-events="none">⏻</text>
                    <!-- Selected colour dot -->
                    ${dotX !== null ? svg`
                        <circle cx="${dotX}" cy="${dotY}" r="3"
                            fill="white" stroke="rgba(0,0,0,0.4)" stroke-width="1"
                            pointer-events="none"/>
                    ` : ''}`;
            }

            default:
                return svg``;
        }
    }

    // ─── Controls below ring ──────────────────────────────────────────────────
    _renderControls() {
        const mode = this.mode || 'brightness';
        const parts = [];

        if (mode === 'color_temp') {
            const min = this.colorTempMin || 2700;
            const max = this.colorTempMax || 6500;
            const ct  = feezal.isEditor ? 4000 : (this._colorTemp ?? min);
            const pct = Math.max(0, Math.min(100, ((ct - min) / (max - min)) * 100));
            parts.push(html`
                <div class="ctrl-label">Color temperature</div>
                <div class="ct-track"
                    @pointerdown="${this._onCtDown}"
                    @pointermove="${this._onCtMove}"
                    @pointerup="${this._onCtUp}">
                    <div class="ct-thumb" style="left:${pct.toFixed(1)}%"></div>
                </div>`);
        }

        if (this.subscribeWhite || this.publishWhite) {
            parts.push(html`
                <div class="ctrl-label">White</div>
                <md-slider min="0" max="100" value="${this._white ?? 0}"
                    @change="${e => this._onWhite(e, this.publishWhite, '_white')}"></md-slider>`);
        }
        if (this.subscribeWarmWhite || this.publishWarmWhite) {
            parts.push(html`
                <div class="ctrl-label">Warm white</div>
                <md-slider min="0" max="100" value="${this._warmWhite ?? 0}"
                    @change="${e => this._onWhite(e, this.publishWarmWhite, '_warmWhite')}"></md-slider>`);
        }
        if (this.subscribeColdWhite || this.publishColdWhite) {
            parts.push(html`
                <div class="ctrl-label">Cold white</div>
                <md-slider min="0" max="100" value="${this._coldWhite ?? 0}"
                    @change="${e => this._onWhite(e, this.publishColdWhite, '_coldWhite')}"></md-slider>`);
        }
        if (this.effects) {
            const list = this.effects.split(',').map(s => s.trim()).filter(Boolean);
            if (list.length > 0) {
                parts.push(html`
                    <md-outlined-select value="${this._effect}" @change="${this._onEffect}">
                        <md-select-option value="">— Effect —</md-select-option>
                        ${list.map(ef => html`<md-select-option value="${ef}">${ef}</md-select-option>`)}
                    </md-outlined-select>`);
            }
        }

        return parts.length > 0 ? html`<div class="controls">${parts}</div>` : '';
    }

    // ─── Render ───────────────────────────────────────────────────────────────
    render() {
        const showUnavail = !feezal.isEditor && this.subscribeAvailability && !this._available;
        return html`
            ${showUnavail ? html`
                <div class="unavail" title="Device unavailable">
                    <svg viewBox="0 0 24 24" fill="currentColor" xmlns="http://www.w3.org/2000/svg">
                        <path d="M24 8.98C20.93 5.9 16.69 4 12 4c-1.69 0-3.32.25-4.86.71l2.5 2.5c.77-.14 1.55-.21 2.36-.21 3.42 0 6.7 1.21 9.32 3.42L24 8.98zM2.81 2.81L1.39 4.22l2.05 2.05C2.2 6.92 1.05 7.86 0 8.98l1.68 1.43c.93-.78 1.94-1.45 3.01-2L6.4 9.83c-1.2.55-2.31 1.3-3.28 2.21L4.81 13.46C5.96 12.38 7.4 11.62 9 11.27l2.16 2.16c-1.3.18-2.5.74-3.46 1.59L12 19.51l1.94-1.94 5.84 5.84 1.41-1.41L2.81 2.81zM12 16.5l-1.41-1.41L12 13.68c.5 0 .96.06 1.42.13l1.71 1.71c-.99-.65-2.18-1.02-3.13-1.02z"/>
                    </svg>
                </div>
            ` : ''}
            <div class="ring-wrap">
                <svg viewBox="0 0 100 100" xmlns="http://www.w3.org/2000/svg"
                    style="cursor:${feezal.isEditor ? 'default' : 'pointer'}"
                    @pointerdown="${this._onSvgPointerDown}">
                    ${this._svgContent()}
                </svg>
            </div>
            ${this._renderControls()}
            ${this.label ? html`<div class="label">${this.label}</div>` : ''}`;
    }
}

customElements.define('feezal-element-material-light', FeezalElementMaterialLight);
export {FeezalElementMaterialLight};

// ─── N6 custom inspector (E35) ──────────────────────────────────────────────
// Replaces the flat 30+ attribute form with a focused two-tab editor. Renders
// only in the editor sidebar (Shoelace components are globally registered
// there); in the viewer this class is defined but never instantiated.
const LIGHT_SECTIONS = [
    {id: 'brightness', title: 'Brightness', topics: [
        {attr: 'subscribe-brightness', label: 'Subscribe'},
        {attr: 'publish-brightness',   label: 'Publish'},
    ]},
    {id: 'color_temp', title: 'Color Temperature', topics: [
        {attr: 'subscribe-color-temp', label: 'Subscribe'},
        {attr: 'publish-color-temp',   label: 'Publish'},
    ]},
    {id: 'color', title: 'Color — RGB / HS', topics: [
        {attr: 'subscribe-rgb', label: 'Subscribe RGB'},
        {attr: 'publish-rgb',   label: 'Publish RGB'},
        {attr: 'subscribe-hs',  label: 'Subscribe HS'},
        {attr: 'publish-hs',    label: 'Publish HS'},
    ]},
    {id: 'white', title: 'White / RGBW / RGBWW', topics: [
        {attr: 'subscribe-white', label: 'White ↓'},
        {attr: 'publish-white',   label: 'White ↑'},
        {attr: 'subscribe-warm-white', label: 'Warm ↓'},
        {attr: 'publish-warm-white',   label: 'Warm ↑'},
        {attr: 'subscribe-cold-white', label: 'Cold ↓'},
        {attr: 'publish-cold-white',   label: 'Cold ↑'},
    ]},
    {id: 'effects', title: 'Effects', topics: [
        {attr: 'subscribe-effect', label: 'Subscribe'},
        {attr: 'publish-effect',   label: 'Publish'},
    ]},
];

// Optional, additive capabilities offered in json mode (no per-property topics
// to derive enablement from, so they are toggled explicitly). Toggling one off
// clears its backing config; toggling on reveals the inline editor.
const JSON_CAPABILITIES = [
    {id: 'effects', title: 'Effects', topics: [
        {attr: 'effects', label: 'Available (comma-separated)', placeholder: 'colorloop, rainbow, …'},
    ]},
];

class FeezalElementMaterialLightInspector extends LitElement {
    static properties = {
        element: {attribute: false},
        _tab:    {state: true},
        _open:   {state: true},
    };

    static styles = css`
        :host { display: block; font-size: 12px; color: var(--feezal-color, #333); }
        sl-tab-panel::part(base) { padding: 8px 2px; }
        .section { border: 1px solid var(--feezal-border, #e0e0e0); border-radius: 6px; margin-bottom: 8px; }
        .sec-head {
            display: flex; align-items: center; gap: 8px;
            padding: 6px 8px; font-weight: 600;
            background: var(--feezal-bg-sub, #f5f5f5);
            border-radius: 6px 6px 0 0;
        }
        .sec-head.collapsed { border-radius: 6px; }
        .sec-title { flex: 1; }
        .sec-body { padding: 8px; display: flex; flex-direction: column; gap: 6px; }
        .field { display: flex; flex-direction: column; gap: 2px; }
        .field > label { font-size: 10px; opacity: 0.6; text-transform: uppercase; letter-spacing: 0.04em; }
        sl-input, sl-select { width: 100%; }
        /* Match the standard attribute/style inspector Shoelace theming
           (input/select background, border and label colours for dark mode). */
        sl-input::part(form-control-label), sl-select::part(form-control-label) { color: var(--sl-input-label-color, inherit); font-size: 12px; }
        sl-input::part(base), sl-select::part(combobox) { background: var(--feezal-bg, #fff); border-color: var(--feezal-border, #ccc); color: var(--feezal-color, #333); }
        sl-input::part(input) { background: var(--feezal-bg, #fff); color: var(--sl-input-color, #333); }
        sl-switch { color: var(--feezal-color, inherit); }
        .row { display: flex; gap: 6px; }
        .row > .field { flex: 1; }
        .hint { font-size: 10px; opacity: 0.55; padding: 0 2px 4px; }
    `;

    constructor() {
        super();
        this.element = null;
        this._tab = 'topics';
        this._open = {};
    }

    willUpdate(changed) {
        if (changed.has('element')) this._open = {};
    }

    // ── attribute access ──────────────────────────────────────────────────
    _val(name) { return this.element?.getAttribute(name) ?? ''; }

    _emit(name, value, rerender = false) {
        this.dispatchEvent(new CustomEvent('feezal-attribute-changed', {
            bubbles: true, composed: true, detail: {name, value},
        }));
        if (rerender) this.requestUpdate();
    }

    _onInput(name, e) { this._emit(name, e.target.value); }
    _onSelect(name, e) { this._emit(name, e.target.value, true); }

    _sectionEnabled(sec) {
        if (this._open[sec.id]) return true;
        return sec.topics.some(t => this._val(t.attr) !== '');
    }

    _toggleSection(sec, e) {
        const on = e.target.checked;
        if (on) {
            this._open = {...this._open, [sec.id]: true};
        } else {
            // Clear all topic attributes for this section
            sec.topics.forEach(t => this._emit(t.attr, ''));
            this._open = {...this._open, [sec.id]: false};
        }
        this.requestUpdate();
    }

    // ── render ────────────────────────────────────────────────────────────
    render() {
        if (!this.element) return html``;
        return html`
            <sl-tab-group @sl-tab-show="${e => { this._tab = e.detail.name; }}">
                <sl-tab slot="nav" panel="topics" ?active="${this._tab === 'topics'}">Topics</sl-tab>
                <sl-tab slot="nav" panel="config" ?active="${this._tab === 'config'}">Config</sl-tab>
                <sl-tab-panel name="topics">${this._renderTopics()}</sl-tab-panel>
                <sl-tab-panel name="config">${this._renderConfig()}</sl-tab-panel>
            </sl-tab-group>
        `;
    }

    _topicInput(t) {
        return html`
            <div class="field">
                <label>${t.label}</label>
                <sl-input size="small" autocomplete="off" placeholder="${t.placeholder ?? 'mqtt/topic'}" value="${this._val(t.attr)}"
                    @sl-change="${e => this._onInput(t.attr, e)}"></sl-input>
            </div>`;
    }

    _renderTopics() {
        // json mode → single State & Control section + optional capabilities
        if (this._val('payload-mode') === 'json') {
            return html`
                <div class="hint">JSON mode — one topic carries the whole state object.</div>
                <div class="section">
                    <div class="sec-head">State &amp; Control</div>
                    <div class="sec-body">
                        ${this._topicInput({attr: 'subscribe', label: 'Subscribe (state)'})}
                        ${this._topicInput({attr: 'publish',   label: 'Publish (…/set)'})}
                    </div>
                </div>
                ${JSON_CAPABILITIES.map(sec => {
                    const enabled = this._sectionEnabled(sec);
                    return html`
                        <div class="section">
                            <div class="sec-head ${enabled ? '' : 'collapsed'}">
                                <span class="sec-title">${sec.title}</span>
                                <sl-switch size="small" ?checked="${enabled}"
                                    @sl-change="${e => this._toggleSection(sec, e)}"></sl-switch>
                            </div>
                            ${enabled ? html`<div class="sec-body">${sec.topics.map(t => this._topicInput(t))}</div>` : ''}
                        </div>`;
                })}`;
        }

        // separate mode → always-on State section + capability-gated sections
        return html`
            <div class="section">
                <div class="sec-head">State</div>
                <div class="sec-body">
                    ${this._topicInput({attr: 'subscribe-state', label: 'Subscribe'})}
                    ${this._topicInput({attr: 'publish-state',   label: 'Publish'})}
                </div>
            </div>
            ${LIGHT_SECTIONS.map(sec => {
                const enabled = this._sectionEnabled(sec);
                return html`
                    <div class="section">
                        <div class="sec-head ${enabled ? '' : 'collapsed'}">
                            <span class="sec-title">${sec.title}</span>
                            <sl-switch size="small" ?checked="${enabled}"
                                @sl-change="${e => this._toggleSection(sec, e)}"></sl-switch>
                        </div>
                        ${enabled ? html`<div class="sec-body">${sec.topics.map(t => this._topicInput(t))}</div>` : ''}
                    </div>`;
            })}`;
    }

    _renderConfig() {
        const isJson = this._val('payload-mode') === 'json';
        const ctEnabled = isJson ||
            this._sectionEnabled(LIGHT_SECTIONS.find(s => s.id === 'color_temp'));
        const brEnabled = isJson ||
            this._sectionEnabled(LIGHT_SECTIONS.find(s => s.id === 'brightness'));
        // In json mode the effect list is toggled from the Topics → Capabilities
        // section, so it is not duplicated here.
        const fxEnabled = !isJson &&
            this._sectionEnabled(LIGHT_SECTIONS.find(s => s.id === 'effects'));

        return html`
            <div class="section">
                <div class="sec-head">Mode</div>
                <div class="sec-body">
                    <div class="field">
                        <label>Active control</label>
                        <sl-select size="small" value="${this._val('mode') || 'brightness'}"
                            @sl-change="${e => this._onSelect('mode', e)}">
                            <sl-option value="brightness">Brightness</sl-option>
                            <sl-option value="color_temp">Color temperature</sl-option>
                            <sl-option value="rgb">RGB</sl-option>
                            <sl-option value="hs">Hue / Saturation</sl-option>
                        </sl-select>
                    </div>
                    <div class="field">
                        <label>Payload mode</label>
                        <sl-select size="small" value="${this._val('payload-mode') || 'separate'}"
                            @sl-change="${e => this._onSelect('payload-mode', e)}">
                            <sl-option value="separate">separate (one topic per property)</sl-option>
                            <sl-option value="json">json (single topic)</sl-option>
                        </sl-select>
                    </div>
                </div>
            </div>

            <div class="section">
                <div class="sec-head">State payloads</div>
                <div class="sec-body row">
                    <div class="field">
                        <label>On</label>
                        <sl-input size="small" autocomplete="off" value="${this._val('payload-on') || 'on'}"
                            @sl-change="${e => this._onInput('payload-on', e)}"></sl-input>
                    </div>
                    <div class="field">
                        <label>Off</label>
                        <sl-input size="small" autocomplete="off" value="${this._val('payload-off') || 'off'}"
                            @sl-change="${e => this._onInput('payload-off', e)}"></sl-input>
                    </div>
                </div>
            </div>

            ${brEnabled ? html`
                <div class="section">
                    <div class="sec-head">Brightness scale</div>
                    <div class="sec-body row">
                        <div class="field">
                            <label>Min</label>
                            <sl-input type="number" size="small" autocomplete="off" value="${this._val('brightness-min') || '0'}"
                                @sl-change="${e => this._onInput('brightness-min', e)}"></sl-input>
                        </div>
                        <div class="field">
                            <label>Max</label>
                            <sl-input type="number" size="small" autocomplete="off" value="${this._val('brightness-max') || '100'}"
                                @sl-change="${e => this._onInput('brightness-max', e)}"></sl-input>
                        </div>
                    </div>
                </div>` : ''}

            ${ctEnabled ? html`
                <div class="section">
                    <div class="sec-head">Color Temperature</div>
                    <div class="sec-body">
                        <div class="field">
                            <label>Unit</label>
                            <sl-select size="small" value="${this._val('color-temp-unit') || 'kelvin'}"
                                @sl-change="${e => this._onSelect('color-temp-unit', e)}">
                                <sl-option value="kelvin">kelvin</sl-option>
                                <sl-option value="mired">mired</sl-option>
                            </sl-select>
                        </div>
                        <div class="row">
                            <div class="field">
                                <label>Min (K)</label>
                                <sl-input type="number" size="small" autocomplete="off" value="${this._val('color-temp-min') || '2700'}"
                                    @sl-change="${e => this._onInput('color-temp-min', e)}"></sl-input>
                            </div>
                            <div class="field">
                                <label>Max (K)</label>
                                <sl-input type="number" size="small" autocomplete="off" value="${this._val('color-temp-max') || '6500'}"
                                    @sl-change="${e => this._onInput('color-temp-max', e)}"></sl-input>
                            </div>
                        </div>
                    </div>
                </div>` : ''}

            ${fxEnabled ? html`
                <div class="section">
                    <div class="sec-head">Effects</div>
                    <div class="sec-body">
                        <div class="field">
                            <label>Available (comma-separated)</label>
                            <sl-input size="small" autocomplete="off" value="${this._val('effects')}"
                                @sl-change="${e => this._onInput('effects', e)}"></sl-input>
                        </div>
                    </div>
                </div>` : ''}

            <div class="section">
                <div class="sec-head">Availability</div>
                <div class="sec-body">
                    <div class="field">
                        <label>Subscribe</label>
                        <sl-input size="small" autocomplete="off" placeholder="…/availability" value="${this._val('subscribe-availability')}"
                            @sl-change="${e => this._onInput('subscribe-availability', e)}"></sl-input>
                    </div>
                    <div class="row">
                        <div class="field">
                            <label>Available</label>
                            <sl-input size="small" autocomplete="off" value="${this._val('payload-available') || 'online'}"
                                @sl-change="${e => this._onInput('payload-available', e)}"></sl-input>
                        </div>
                        <div class="field">
                            <label>Unavailable</label>
                            <sl-input size="small" autocomplete="off" value="${this._val('payload-unavailable') || 'offline'}"
                                @sl-change="${e => this._onInput('payload-unavailable', e)}"></sl-input>
                        </div>
                    </div>
                </div>
            </div>

            <div class="section">
                <div class="sec-head">Display</div>
                <div class="sec-body">
                    <div class="field">
                        <label>Label</label>
                        <sl-input size="small" autocomplete="off" value="${this._val('label')}"
                            @sl-change="${e => this._onInput('label', e)}"></sl-input>
                    </div>
                </div>
            </div>`;
    }
}

customElements.define('feezal-element-material-light-inspector', FeezalElementMaterialLightInspector);
export {FeezalElementMaterialLightInspector};
