// @feezal/feezal-element-paper-dialog/feezal-element-paper-dialog.js
import {FeezalElement, html} from "@feezal/feezal-element";

import '@polymer/paper-dialog/paper-dialog';


class FeezalElementPaperDialog extends FeezalElement {
    static get template() {
        return html`
            <style include="feezal-style-element"></style>
            <div>
                <paper-dialog>
                    <h2>Header</h2>
                    <paper-dialog-scrollable>
                        Lorem ipsum...
                    </paper-dialog-scrollable>
                    <div class="buttons">
                        <paper-button dialog-dismiss>Cancel</paper-button>
                        <paper-button dialog-confirm autofocus>Accept</paper-button>
                    </div>
                </paper-dialog>
            </div>
        `;
    }
    static get properties() {
        return {
            subscribeTopic: {
                type: String,
                value: '',
                reflectToAttribute: true,
            }


        }
    }


    static get feezal() {
        return {
            palette: {
                category: 'Paper',
                name: 'Dialog',
                color: '#eee'
            },
            attributes: [

            ],
            styles: [
                'top',
                'left',
                'width',
                'height',
                'font',
                'color',
                'text-align',
                'background',
                'border',
                'overflow',
            ],
            restrict: {
                minWidth: 36,
                minHeight: 12
            },
            defaultStyle: {
                width: '60px',
                height: '20px'
            }
        };
    }


}

window.customElements.define('feezal-element-paper-dialog', FeezalElementPaperDialog);

export {FeezalElementPaperDialog};
