// @feezal/feezal-element-basic-datetime/feezal-element-basic-datetime.js
import {FeezalElement, html} from "@feezal/feezal-element";

import {format} from 'date-fns';

class FeezalElementBasicDatetime extends FeezalElement {
    static get template() {
        return html`
            <style include="feezal-style-element"></style>
            <span>[[formatedValue]]</span>
        `;
    }
    static get properties() {
        return {
            subscribeTopic: {
                type: String,
                value: '',
                reflectToAttribute: true,
            },
            format: {
                type: String,
                reflectToAttribute: true,
                value: 'dd.MM.yyyy hh:mm:ss'
            }
        }
    }


    static get feezal() {
        return {
            palette: {
                category: 'Basic',
                name: 'Datetime',
                color: '#eee'
            },
            attributes: [
                'digits',
                'decimalSeparator',
                'prefix',
                'suffix'

            ],
            baseAttribute: 'value',
            styles: [
                'top',
                'left',
                'width',
                'height',
                'font',
                'color',
                'text-align',
                'background',
                'border',
                'overflow',
            ],
            restrict: {
                minWidth: 36,
                minHeight: 12
            },
            defaultStyle: {
                width: '60px',
                height: '20px'
            }
        };
    }

    connectedCallback() {
        super.connectedCallback();
        this.interval = setInterval(this._update.bind(this), 1000);
    }

    disconnectedCallback() {
        super.disconnectedCallback();
        clearInterval(this.interval);
    }

    _update() {
        const now = new Date();
        this.formatedValue = format(now, this.format);
    }

}

window.customElements.define('feezal-element-basic-datetime', FeezalElementBasicDatetime);

export {FeezalElementBasicDatetime};
