// @feezal/feezal-element-basic-datetime/feezal-element-basic-datetime.js
import {FeezalElement, html} from "@feezal/feezal-element";

import {zonedTimeToUtc, utcToZonedTime, format} from 'date-fns-tz';

import {be, de, enGB, enUS, es, fr, hr, hu, it, nl, pl, ru} from 'date-fns/locale'
const locales = {be, de, enGB, enUS, es, fr, hr, hu, it, nl, pl, ru};

class FeezalElementBasicDatetime extends FeezalElement {
    static get template() {
        return html`
            <style include="feezal-style-element"></style>
            <span>[[formatedValue]]</span>
        `;
    }
    static get properties() {
        return {
            subscribeTopic: {
                type: String,
                value: '',
                reflectToAttribute: true,
            },
            format: {
                type: String,
                reflectToAttribute: true,
                value: 'dd.MM.yyyy hh:mm:ss'
            },
            locale: {
                type: String,
                reflectToAttribute: true,
                value: 'enUS'
            },
            timezone: {
                type: String,
                reflectToAttribute: true,
                value: Intl.DateTimeFormat().resolvedOptions().timeZone
            }
        }
    }


    static get feezal() {
        return {
            palette: {
                category: 'Basic',
                name: 'Datetime',
                color: '#eee'
            },
            attributes: [
                {
                    name: 'format' // https://date-fns.org/v2.8.1/docs/format
                },
                {
                    name: 'locale',
                    type: 'dropdown',
                    options: Object.keys(locales)
                },
                {
                    name: 'timezone'
                }
            ],
            baseAttribute: 'value',
            styles: [
                'top',
                'left',
                'width',
                'height',
                'font',
                'color',
                'text-align',
                'background',
                'border',
                'overflow',
            ],
            restrict: {
                minWidth: 30,
                minHeight: 12
            },
            defaultStyle: {
                width: '120px',
                height: '24px'
            }
        };
    }

    connectedCallback() {
        super.connectedCallback();
        this.interval = setInterval(this._update.bind(this), 1000);
    }

    disconnectedCallback() {
        super.disconnectedCallback();
        clearInterval(this.interval);
    }

    _update() {
        const now = new Date();
        const locale = locales[this.locale];
        this.formatedValue = format(now, this.format, {locale});
    }

}

window.customElements.define('feezal-element-basic-datetime', FeezalElementBasicDatetime);

export {FeezalElementBasicDatetime};
