// @feezal/feezal-element-paper-checkbox/feezal-element-paper-checkbox.js
import {FeezalElement, html} from "@feezal/feezal-element";

import '@polymer/paper-checkbox';

class FeezalElementPaperCheckbox extends FeezalElement {
    static get template() {
        return html`
            <style include="feezal-style-element"></style>
            <style>
                :host {
                    overflow: visible;
                }
                paper-checkbox {
                    width: 100%;
                    height: 100%;
                }
            </style>
            <paper-checkbox 
                noink="[[noink]]" 
                disabled="[[disabled]]" 
                invalid="[[invalid]]"
                checked="[[checked]]"
                id="checkbox"
            >[[label]]</paper-checkbox>
        `;
    }
    static get properties() {
        return {
            topic: {
                type: String,
                    value: '',
                    reflectToAttribute: true
            },
            label: {
                type: String,
                value: '',
                reflectToAttribute: true
            },
            disabled: {
                type: Boolean,
                value: false,
                reflectToAttribute: true
            },
            invalid: {
                type: Boolean,
                value: false,
                reflectToAttribute: true
            },
            checked: {
                type: Boolean,
                reflectToAttribute: true
            },
            noink: {
                type: Boolean,
                value: false,
                reflectToAttribute: true
            }
        }
    }
    static get feezal() {
        return {
            palette: {
                category: 'Paper',
                    name: 'Checkbox',
                    color: '#eee'
            },
            baseAttribute: 'checked',
            styles: [
                'top',
                'left',
                'width',
                'height',
                '--paper-checkbox-size'
            ],
            attributes: ['noink'],
            restrict: {
                minWidth: 40,
                minHeight: 22
            }
        }

    }

    connectedCallback() {
        super.connectedCallback();
        feezal.connection.subscribe(this.topic, msg => {
            if (msg.payload) {
                this.setAttribute('checked', true);
            } else {
                this.removeAttribute('checked');
            }
        });
        this.$.checkbox.addEventListener('checked-changed', event => {
            feezal.connection.publish(this.topic, event.detail.value);
        });
    }
}

window.customElements.define('feezal-element-paper-checkbox', FeezalElementPaperCheckbox);

export {FeezalElementPaperCheckbox};
