// @feezal/feezal-element-paper-dropdown/feezal-element-paper-dropdown.js
import {FeezalElement, html} from "@feezal/feezal-element";

class FeezalElementPaperDropdown extends FeezalElement {
    static get template() {
        return html`
            <style include="feezal-style-element"></style>
            <paper-dropdown-menu label="[[label]]" value="[[value]]" on-value-changed="_change">
                <paper-listbox slot="dropdown-content">
                     <template is="dom-repeat" items="[[items]]">
                        <paper-item>[[item]]</paper-item>
                     </template>
                </paper-listbox>
            </paper-dropdown-menu>
        `;
    }
    static get properties() {
        return {
            subscribeTopic: {
                type: String,
                value: '',
                reflectToAttribute: true,
            },
            label: {
                type: String,
                value: '',
                reflectToAttribute: true,
            },
            items: {
                type: Array,
                reflectToAttribute: true
            },
            value: {
                type: String,
                value: '',
                reflectToAttribute: true,
                notify: true
            }
        }
    }


    static get feezal() {
        return {
            palette: {
                category: 'Paper',
                name: 'Dropdown',
                color: '#eee'
            },
            attributes: [

            ],
            styles: [
                'top',
                'left',
                'width',
                'height',
                'font',
                'color',
                'text-align',
                'background',
                'border',
                'overflow',
            ],
            restrict: {
                minWidth: 36,
                minHeight: 12
            },
            defaultStyle: {
                width: '60px',
                height: '20px'
            }
        };
    }

    _change(value) {

    }
    _deserializeValue(value, type) {
        if (type === Object || type === Array) {
            return JSON.parse(value.replace(/&quot;/g, '"'));
        } else {
            return super._deserializeValue(value, type);
        }
    }

    _serializeValue(value) {
        if (typeof value === 'object') {
            return JSON.stringify(value).replace(/"/g, '&quot;');
        }
        return super._serializeValue(value);
    }

}

window.customElements.define('feezal-element-paper-dropdown', FeezalElementPaperDropdown);

export {FeezalElementPaperDropdown};
