// @feezal/feezal-element-paper-dropdown/feezal-element-paper-dropdown.js
import {FeezalElement, html} from "@feezal/feezal-element";

class FeezalElementPaperDropdown extends FeezalElement {
    static get template() {
        return html`
            
        `;
    }
    static get properties() {
        return {
            subscribeTopic: {
                type: String,
                value: '',
                reflectToAttribute: true,
            }


        }
    }


    static get feezal() {
        return {
            palette: {
                category: 'Paper',
                name: 'Dropdown',
                color: '#eee'
            },
            attributes: [

            ],
            styles: [
                'top',
                'left',
                'width',
                'height',
                'font',
                'color',
                'text-align',
                'background',
                'border',
                'overflow',
            ],
            restrict: {
                minWidth: 36,
                minHeight: 12
            },
            defaultStyle: {
                width: '60px',
                height: '20px'
            }
        };
    }


}

window.customElements.define('feezal-element-paper-dropdown', FeezalElementPaperDropdown);

export {FeezalElementPaperDropdown};
