// @feezal/feezal-element-paper-card-template/feezal-element-paper-card-template.js
import {FeezalElement, html} from "@feezal/feezal-element";
import '@polymer/paper-card';

class FeezalElementPaperCardTemplate extends FeezalElement {
    static get template() {
        return html`
            <style include="feezal-style-element"></style>
            <style>
                :host {
                    overflow: visible;
                    --content-overflow: scroll;
                }
                paper-card {
                    width: 100%;
                    height: 100%;
                    overflow: hidden;
                }
                .card-content {
                    overflow: var(--content-overflow);
                }
                .card-actions {
                    display: none;
                }
            </style>
            <paper-card 
                heading="[[heading]]" 
                image="[[image]]" 
                elevation="[[elevation]]"
                animated-shadow="[[animatedShadow]]" 
            >
                <div id="content" class="card-content"></div>
                <div id="actions" class="card-actions"></div>
            </paper-card>
        `;
    }
    static get properties() {
        return {
            topic: {
                type: String,
                value: '',
                reflectToAttribute: true,
                observer: '_topicChanged'
            },
            actions: {
                type: String,
                value: '',
                reflectToAttribute: true,
                observer: '_actionsChanged',
                feezal: {
                    textarea: true
                }
            },
            msg: {
                type: Object,
                value: {},
                observer: '_msgChanged'
            },
            heading: {
                type: String,
                value: '',
                reflectToAttribute: true
            },
            image: {
                type: String,
                value: '',
                reflectToAttribute: true
            },
            elevation: {
                type: Number,
                value: 1,
                reflectToAttribute: true,
                feezal: {
                    input: true,
                    inputType: 'number',
                    min: 0,
                    max: 5,
                    step: 1
                }
            },
            animatedShadow: {
                type: Boolean,
                value: true,
                reflectToAttribute: true
            }
        }
    }

    static get feezal() {
        return {
            palette: {
                category: 'Paper',
                name: 'Template',
                color: '#eee'
            },

            attributes: [
                'topic',
                'heading',
                {name: 'elevation', type: Number, min: 0, max: 5, step: 1},
                {name: 'animatedShadow', type: Boolean},
                {name: 'image', image: true},
                {name: 'template', textarea: true, template: true}
            ],
            styles: [
                'top',
                'left',
                'width',
                'height',
                'font',
                'color',
                'text-align',
                '--paper-card-background-color',
                '--paper-card-header-color',
                '--content-overflow',
            ],


            restrict: {
                minWidth: 12,
                minHeight: 12
            },
            defaultStyle: {
                width: '200px',
                height: '120px',
                color: 'var(--primary-text-color)'
            }
        }
    }
    connectedCallback() {
        super.connectedCallback();
        feezal.connection.subscribe(this.topic, msg => {
            this.msg = msg;
        });
    }

    _msgChanged() {
        if (!this._processTemplate && this.querySelector('template')) {
            // evil eval...
            this._processTemplate = new Function('msg', "return `" + this.querySelector('template').innerHTML + "`;");
        }

        if (!this.msg || Object.keys(this.msg).length === 0) {
            return;
        }
        try {
            this.$.content.innerHTML = this._processTemplate(this.msg);
        } catch (error) {
            console.error(error.message);
        }
    }

    _topicChanged() {
        if (this.subscription) {
            feezal.connection.unsubscribe(this.subscription);
        }
        if (this.topic) {
            this.subscription = feezal.connection.subscribe(this.topic, msg => this.msg = msg);
        }
    }

    _actionsChanged() {
        if (this.actions) {
            this.$.actions.style.display = 'block';
        } else {
            this.$.actions.style.display = 'none';
        }
    }

}

window.customElements.define('feezal-element-paper-card-template', FeezalElementPaperCardTemplate);

export {FeezalElementPaperCardTemplate};
