// @feezal/feezal-element-paper-card-template/feezal-element-paper-card-template.js
import {FeezalElement, html} from "@feezal/feezal-element";
import '@polymer/paper-card';

class FeezalElementPaperCardTemplate extends FeezalElement {
    static get template() {
        return html`
            <style include="feezal-style-element"></style>
            <style>
                :host {
                    overflow: visible;
                    --content-overflow: scroll;
                }
                paper-card {
                    width: 100%;
                    height: 100%;
                    overflow: hidden;
                }
                .card-content {
                    overflow: var(--content-overflow);
                }
                .card-actions {
                    display: none;
                }
            </style>
            <paper-card 
                heading="[[heading]]" 
                image="[[image]]" 
                elevation="[[elevation]]"
                animated-shadow="[[animatedShadow]]" 
            >
                <div id="content" class="card-content"></div>
                <div id="actions" class="card-actions"></div>
            </paper-card>
        `;
    }
    static get properties() {
        return {
            topic: {
                type: String,
                value: '',
                reflectToAttribute: true,
                observer: '_topicChanged'
            },
            content: {
                type: String,
                value: '',
                reflectToAttribute: true,
                observer: '_contentChanged',
                feezal: {
                    textarea: true
                }
            },
            actions: {
                type: String,
                value: '',
                reflectToAttribute: true,
                observer: '_actionsChanged',
                feezal: {
                    textarea: true
                }
            },
            hideUndefined: {
                type: Boolean,
                value: false,
                reflectToAttribute: true,
                observer: '_contentChanged',
            },
            msg: {
                type: Object,
                value: {},
                observer: '_contentChanged'
            },
            heading: {
                type: String,
                value: '',
                reflectToAttribute: true
            },
            image: {
                type: String,
                value: '',
                reflectToAttribute: true
            },
            elevation: {
                type: Number,
                value: 1,
                reflectToAttribute: true,
                feezal: {
                    input: true,
                    inputType: 'number',
                    min: 0,
                    max: 5,
                    step: 1
                }
            },
            animatedShadow: {
                type: Boolean,
                value: true,
                reflectToAttribute: true
            }
        }
    }
    _processTemplate(template, vars = {}) {
        const match = template.match(/\${[^}]+}/g);
        if (match) {
            match.forEach(v => {
                const key = v.substr(2, v.length - 3);
                const rx = new RegExp('\\${' + key + '}', 'g');
                let replacement;
                if (this.hideUndefined && typeof vars[key] === 'undefined') {
                    replacement = '';
                } else {
                    replacement = vars[key];
                }
                template = template.replace(rx, replacement);
            });
        }
        return template;
    }
    _contentChanged() {
        this.$.content.innerHTML = this._processTemplate(this.content, this.msg);
    }
    _actionsChanged() {
        if (this.actions) {
            this.$.actions.style.display = 'block';
        } else {
            this.$.actions.style.display = 'none';
        }
    }

    static get feezal() {
        return {
            palette: {
                category: 'Paper',
                name: 'Template',
                color: '#eee'
            },
            styles: [
                'top',
                'left',
                'width',
                'height',
                'font',
                'color',
                'text-align',
                '--paper-card-background-color',
                '--paper-card-header-color',
                '--content-overflow',
            ],


            restrict: {
                minWidth: 12,
                minHeight: 12
            },
            defaultStyle: {
                width: '200px',
                height: '120px'
            }
        }
    }
    connectedCallback() {
        super.connectedCallback();
        feezal.connection.subscribe(this.topic, msg => {
            this.msg = msg;
        });
    }
    _topicChanged() {

    }
}

window.customElements.define('feezal-element-paper-card-template', FeezalElementPaperCardTemplate);

export {FeezalElementPaperCardTemplate};
