// @feezal/feezal-element-basic-iframe/feezal-element-basic-iframe.js
import {FeezalElement, html} from "@feezal/feezal-element";

class FeezalElementBasicIframe extends FeezalElement {
    static get template() {
        return html`
            <style include="feezal-style-element">
                iframe {
                    border: var(--feezal-basic-iframe-border);
                    padding: 0;
                    margin: 0;
                    width: 100%;
                    height: 100%;
                    box-sizing: border-box;
                }
            </style>
            <iframe src="[[src]]" frameborder="0"></iframe>
        `;
    }
    static get properties() {
        return {
            topic: {
                type: String,
                value: '',
                reflectToAttribute: true
            },
            src: {
                type: String,
                value: '',
                reflectToAttribute: true
            },
            persist: {
                type: Boolean,
                reflectToAttribute: true,
                feezal: {
                    checkbox: true
                }
            },
            preload: {
                type: Boolean,
                reflectToAttribute: true,
                feezal: {
                    checkbox: true
                }
            },
            refresh: {
                type: Number,
                value: 0,
                reflectToAttribute: true
            },
        }
    }
    static get feezal() {
        return {
            palette: {
                category: 'Basic',
                name: 'iframe',
                color: '#eee'
            },
            attributes: [
                'src',
                'persist',
                'preload',
                'refresh'
            ],
            baseAttribute: 'src',
            styles: [
                'top',
                'left',
                'width',
                'height',
                '--feezal-basic-iframe-border'
            ],
            restrict: {
                minWidth: 12,
                minHeight: 12
            }
        };
    }
    connectedCallback() {
        super.connectedCallback();
    }
}

window.customElements.define('feezal-element-basic-iframe', FeezalElementBasicIframe);

export {FeezalElementBasicIframe};
